File: app/src/main/java/de/grobox/transportr/data/locations/LocationRepository.java
package de.grobox.transportr.data.locations;


import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Transformations;
import android.support.annotation.Nullable;
import android.support.annotation.WorkerThread;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import de.grobox.transportr.AbstractManager;
import de.grobox.transportr.data.locations.FavoriteLocation.FavLocationType;
import de.grobox.transportr.locations.WrapLocation;
import de.grobox.transportr.networks.TransportNetworkManager;
import de.schildbach.pte.NetworkId;

import static de.schildbach.pte.dto.LocationType.COORD;

@Singleton
public class LocationRepository extends AbstractManager {

	private final LocationDao locationDao;

	private final LiveData<NetworkId> networkId;
	private final LiveData<HomeLocation> home;
	private final LiveData<WorkLocation> work;
	private final LiveData<List<FavoriteLocation>> locations;

	@Inject
	public LocationRepository(LocationDao locationDao, TransportNetworkManager transportNetworkManager) {
		this.locationDao = locationDao;
		this.networkId = transportNetworkManager.getNetworkId();
		this.home = Transformations.switchMap(networkId, this::getHomeLocation);
		this.work = Transformations.switchMap(networkId, this::getWorkLocation);
		this.locations = Transformations.switchMap(networkId, this::getFavoriteLocations);
	}

	private LiveData<HomeLocation> getHomeLocation(NetworkId id) {
		return locationDao.getHomeLocation(id);
	}

	public LiveData<HomeLocation> getHomeLocation() {
		return home;
	}

	public void setHomeLocation(WrapLocation location) {
		if (networkId.getValue() == null) return;

		runOnBackgroundThread(() -> {
			// add also as favorite location if it doesn't exist already
			FavoriteLocation favoriteLocation = getFavoriteLocation(networkId.getValue(), location);
			if (favoriteLocation == null) locationDao.addFavoriteLocation(new FavoriteLocation(networkId.getValue(), location));

			locationDao.addHomeLocation(new HomeLocation(networkId.getValue(), location));
		});
	}

	private LiveData<WorkLocation> getWorkLocation(NetworkId id) {
		return locationDao.getWorkLocation(id);
	}

	public LiveData<WorkLocation> getWorkLocation() {
		return work;
	}

	public void setWorkLocation(WrapLocation location) {
		if (networkId.getValue() == null) return;

		runOnBackgroundThread(() -> {
			// add also as favorite location if it doesn't exist already
			FavoriteLocation favoriteLocation = getFavoriteLocation(networkId.getValue(), location);
			if (favoriteLocation == null) locationDao.addFavoriteLocation(new FavoriteLocation(networkId.getValue(), location));

			locationDao.addWorkLocation(new WorkLocation(networkId.getValue(), location));
		});
	}

	public LiveData<List<FavoriteLocation>> getFavoriteLocations() {
		return locations;
	}

	public void addFavoriteLocation(WrapLocation wrapLocation, FavLocationType type) {
		if (wrapLocation.type == COORD) return;

		FavoriteLocation favoriteLocation;
		if (wrapLocation instanceof FavoriteLocation) {
			favoriteLocation = (FavoriteLocation) wrapLocation;
			favoriteLocation.add(type);
		} else if (!(wrapLocation instanceof StoredLocation)) {  // no home or work location please
			favoriteLocation = new FavoriteLocation(networkId.getValue(), wrapLocation);
		} else {
			// nothing for us to do here
			return;
		}
		runOnBackgroundThread(() -> {
			FavoriteLocation locationToStore = favoriteLocation;
			if (favoriteLocation.getUid() == 0) {
				FavoriteLocation existingLocation = getFavoriteLocation(networkId.getValue(), wrapLocation);
				if (existingLocation != null) {
					locationToStore = existingLocation;
					locationToStore.add(type);
				}
			}
			locationDao.addFavoriteLocation(locationToStore);
		});
	}

	@WorkerThread
	private LiveData<List<FavoriteLocation>> getFavoriteLocations(NetworkId networkId) {
		return locationDao.getFavoriteLocations(networkId);
	}

	@Nullable
	@WorkerThread
	private FavoriteLocation getFavoriteLocation(NetworkId networkId, @Nullable WrapLocation l) {
		if (l == null) return null;
		return locationDao.getFavoriteLocation(networkId, l.type, l.id, l.lat, l.lon, l.place, l.name);
	}

}

File: app/src/main/java/de/grobox/transportr/data/locations/LocationRepository.kt
/*
 *    Transportr
 *
 *    Copyright (c) 2013 - 2021 Torsten Grote
 *
 *    This program is Free Software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as
 *    published by the Free Software Foundation, either version 3 of the
 *    License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.grobox.transportr.data.locations


import androidx.lifecycle.LiveData
import androidx.lifecycle.Transformations
import androidx.annotation.WorkerThread
import de.grobox.transportr.AbstractManager
import de.grobox.transportr.data.locations.FavoriteLocation.FavLocationType
import de.grobox.transportr.locations.WrapLocation
import de.grobox.transportr.locations.WrapLocation.WrapType.NORMAL
import de.grobox.transportr.networks.TransportNetworkManager
import de.schildbach.pte.NetworkId
import de.schildbach.pte.dto.LocationType.ADDRESS
import de.schildbach.pte.dto.LocationType.COORD
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class LocationRepository @Inject
constructor(private val locationDao: LocationDao, transportNetworkManager: TransportNetworkManager) : AbstractManager() {

    private val networkId: LiveData<NetworkId> = transportNetworkManager.networkId
    val homeLocation: LiveData<HomeLocation> = Transformations.switchMap(networkId, locationDao::getHomeLocation)
    val workLocation: LiveData<WorkLocation> = Transformations.switchMap(networkId, locationDao::getWorkLocation)
    val favoriteLocations: LiveData<List<FavoriteLocation>> = Transformations.switchMap(networkId, locationDao::getFavoriteLocations)

    fun setHomeLocation(location: WrapLocation) {
        runOnBackgroundThread {
            // add also as favorite location if it doesn't exist already
            val favoriteLocation = getFavoriteLocation(networkId.value, location)
            if (favoriteLocation == null) locationDao.addFavoriteLocation(FavoriteLocation(networkId.value, location))

            locationDao.addHomeLocation(HomeLocation(networkId.value!!, location))
        }
    }

    fun setWorkLocation(location: WrapLocation) {
        runOnBackgroundThread {
            // add also as favorite location if it doesn't exist already
            val favoriteLocation = getFavoriteLocation(networkId.value, location)
            if (favoriteLocation == null) locationDao.addFavoriteLocation(FavoriteLocation(networkId.value, location))

            locationDao.addWorkLocation(WorkLocation(networkId.value!!, location))
        }
    }

    @WorkerThread
    fun addFavoriteLocation(wrapLocation: WrapLocation, type: FavLocationType): FavoriteLocation? {
        if (wrapLocation.type == COORD || wrapLocation.wrapType != NORMAL) return null

        val favoriteLocation = if (wrapLocation is FavoriteLocation) {
            wrapLocation
        } else {
            val location = findExistingLocation(wrapLocation)
            location as? FavoriteLocation ?: FavoriteLocation(networkId.value, wrapLocation)
        }
        favoriteLocation.add(type)

        return if (favoriteLocation.uid != 0L) {
            locationDao.addFavoriteLocation(favoriteLocation)
            favoriteLocation
        } else {
            val existingLocation = getFavoriteLocation(networkId.value, favoriteLocation)
            if (existingLocation != null) {
                existingLocation.add(type)
                locationDao.addFavoriteLocation(existingLocation)
                existingLocation
            } else {
                val uid = locationDao.addFavoriteLocation(favoriteLocation)
                FavoriteLocation(uid, networkId.value, favoriteLocation)
            }
        }
    }

    @WorkerThread
    private fun getFavoriteLocation(networkId: NetworkId?, l: WrapLocation?): FavoriteLocation? {
        return if (l == null) null else locationDao.getFavoriteLocation(networkId, l.type, l.id, l.lat, l.lon, l.place, l.name)
    }

    /**
     * This checks existing ADDRESS Locations for geographic proximity
     * before adding the given location as a favorite.
     * The idea is too prevent duplicates of addresses with slightly different coordinates.
     *
     * @return The given {@link WrapLocation} or if found, the existing one
     */
    private fun findExistingLocation(location: WrapLocation): WrapLocation {
        favoriteLocations.value?.filter {
            it.type == ADDRESS && it.name != null && it.name == location.name && it.isSamePlace(location)
        }?.forEach { return it }
        return location
    }

}
